extmod/modussl_mbedtls.c

```diff
@@ -77,33 +77,52 @@ int null_entropy_func(void *data, unsigned char *output, size_t len) {
 }
 
 int _mbedtls_ssl_send(void *ctx, const byte *buf, size_t len) {
-    mp_obj_t sock = *(mp_obj_t*)ctx;
+    mp_obj_ssl_socket_t *o = (mp_obj_ssl_socket_t*)ctx;
+    mp_obj_t sock = o->sock;
 
     const mp_stream_p_t *sock_stream = mp_get_stream_raise(sock, MP_STREAM_OP_WRITE);
     int err;
 
     int out_sz = sock_stream->write(sock, buf, len, &err);
     if (out_sz == MP_STREAM_ERROR) {
+        if (mp_is_nonblocking_error(err)) {
+            return MBEDTLS_ERR_SSL_WANT_WRITE;
+        }
         return -err;
-    } else {
-        return out_sz;
     }
+    return out_sz;
 }
 
 int _mbedtls_ssl_recv(void *ctx, byte *buf, size_t len) {
-    mp_obj_t sock = *(mp_obj_t*)ctx;
+    mp_obj_ssl_socket_t *o = (mp_obj_ssl_socket_t*)ctx;
+    mp_obj_t sock = o->sock;
 
     const mp_stream_p_t *sock_stream = mp_get_stream_raise(sock, MP_STREAM_OP_READ);
     int err;
-
     int out_sz = sock_stream->read(sock, buf, len, &err);
     if (out_sz == MP_STREAM_ERROR) {
+        if (mp_is_nonblocking_error(err)) {
+            return MBEDTLS_ERR_SSL_WANT_READ;
+        }
         return -err;
-    } else {
-        return out_sz;
     }
+    return out_sz;
 }
 
+STATIC mp_obj_t socket_setblocking(mp_obj_t self_in, mp_obj_t flag_in) {
```

Comment:
The choice which produces least changes (in terms of diff lines) is usually the right one (there're exceptions, but they should be just that - exceptions). Hence, no. 3 ("Put a function declaration at the top of the file").

But the point of minimizing unrelated changes is to actually concentrate on the functional changes introduced by a patch. So:

> This is now called from other functions

Why is it now called from other functions? And go further a bit, I don't think it should be called from other functions. .setblocking() is purely application-level function, which should be called by a user app, not otherwise.